lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,470 @@
+    meta = patch_meta_line(meta, state.parent_meta)
+
+    left = Enum.map(left, &do_normalize(&1, %{state | parent_meta: meta}))
+    right = do_normalize(right, %{state | parent_meta: meta})
+    {:->, meta, [left, right]}
+  end
+
+  # Maps
+  defp do_normalize({:%{}, meta, args}, state) do
+    meta =
+      if meta == [] do
+        line = state.parent_meta[:line]
+        [line: line, closing: [line: line]]
+      else
+        meta
+      end
+
+    args =
+      case args do
+        [{:|, pipe_meta, [left, right]}] ->
+          left = do_normalize(left, meta)
+          right = normalize_args(right, %{state | parent_meta: meta})
+          [{:|, pipe_meta, [left, right]}]
+
+        [{_, _, _} = call] ->
+          [do_normalize(call, %{state | parent_meta: meta})]
+
+        args ->
+          normalize_map_args(args, %{state | parent_meta: meta})
+      end
+
+    {:%{}, meta, args}
+  end
+
+  # Sigils
+  defp do_normalize({sigil, meta, [{:<<>>, _, _} = string, modifiers]} = quoted, state)
+       when is_atom(sigil) do
+    case Atom.to_string(sigil) do
+      <<"sigil_", _name>> ->
+        meta = patch_meta_line(meta, state.parent_meta)
+
+        {sigil, meta, [do_normalize(string, %{state | parent_meta: meta}), modifiers]}
+
+      _ ->
+        normalize_call(quoted, state)
+    end
+  end
+
+  # Calls
+  defp do_normalize({_, _, args} = quoted, state) when is_list(args) do
+    normalize_call(quoted, state)
+  end
+
+  # Integers, floats, atoms
+  defp do_normalize(literal, state) when is_literal(literal) do
+    meta = [line: state.parent_meta[:line]]
+
+    meta =
+      if is_integer(literal) or is_float(literal) do
+        Keyword.put(meta, :token, inspect(literal))
+      else
+        meta
+      end
+
+    meta =
+      if not is_nil(state.parent_meta[:format]) do
+        Keyword.put(meta, :format, state.parent_meta[:format])
+      else
+        meta
+      end
+
+    literal = maybe_escape_literal(literal, state)
+
+    if module_atom?(literal) do
+      "Elixir." <> segments = Atom.to_string(literal)
+
+      segments =
+        segments
+        |> String.split(".")
+        |> Enum.map(&String.to_atom/1)
+
+      {:__aliases__, meta, segments}
+    else
+      {:__block__, meta, [literal]}
+    end
+  end
+
+  # 2-tuples
+  defp do_normalize({left, right}, state) do
+    meta = [line: state.parent_meta[:line]]
+
+    left_state =
+      if is_atom(left) and not module_atom?(left) do
+        %{state | parent_meta: Keyword.put(state.parent_meta, :format, :keyword)}
+      else
+        %{state | parent_meta: meta}
+      end
+
+    {:__block__, meta,
+     [
+       {do_normalize(left, left_state), do_normalize(right, state)}
+     ]}
+  end
+
+  # Lists
+  defp do_normalize(list, state) when is_list(list) do
+    if list != [] and List.ascii_printable?(list) do
+      # It's a charlist
+      list =
+        if state.escape do
+          {string, _} = Code.Identifier.escape(IO.chardata_to_string(list), -1)
```

Comment:
Nice trick with -1 and it makes sense too. The formatter will escape the escape character after all.